eval/src/vespa/eval/eval/simple_value.h

```diff
@@ -0,0 +1,215 @@
+            // create_view. Results from the lookup is extracted using
+            // the next_result function.
+            virtual void lookup(const std::vector<const vespalib::stringref*> &addr) = 0;
+
+            // Extract the next result (if any) from the previous
+            // lookup into the given partial address and index. Only
+            // the labels for the dimensions NOT specified in
+            // create_view will be extracted here.
+            virtual bool next_result(const std::vector<vespalib::stringref*> &addr_out, size_t &idx_out) = 0;
+
+            virtual ~View() {}
+        };
+
+        // total number of mappings (equal to the number of dense subspaces)
+        virtual size_t size() const = 0;
+
+        // create a view able to look up dense subspaces based on
+        // labels from a subset of the mapped dimensions.
+        virtual std::unique_ptr<View> create_view(const std::vector<size_t> &dims) const = 0;
+
+        virtual ~Index() {}
+    };
+    virtual TypedCells cells() const = 0;
+    virtual const Index &index() const = 0;
+    virtual ~NewValue() {}
+};
+
+/**
+ * Tagging interface used as return type from factories before
+ * downcasting to actual builder with specialized cell type.
+ **/
+struct ValueBuilderBase {
+    virtual ~ValueBuilderBase() {}
+};
+
+/**
+ * Interface used to build a value one dense subspace at a
+ * time. Enables decoupling of what the value should contain from how
+ * to store the value.
+ **/
+template <typename T>
+struct ValueBuilder : ValueBuilderBase {
+    // add a dense subspace for the given address (label for all
+    // mapped dimensions in canonical order). Note that previously
+    // returned subspaces will be invalidated when new subspaces are
+    // added. Also note that adding the same subspace multiple times
+    // is not allowed.
+    virtual ArrayRef<T> add_subspace(const std::vector<vespalib::stringref> &addr) = 0;
+
+    // Given the ownership of the builder itself, produce the newly
+    // created value. This means that builders can only be used once,
+    // it also means values can build themselves.
+    virtual std::unique_ptr<NewValue> build(std::unique_ptr<ValueBuilder> self) = 0;
+};
+
+/**
+ * Factory able to create appropriate value builders. We do not really
+ * care about the full mathematical type here, but it needs to be
+ * passed since it is exposed in the value api. The expected number of
+ * subspaces is also passed since it enables the builder to pre-size
+ * internal structures appropriately. Note that since we are not able
+ * to have virtual templated functions we need to cast the created
+ * builder. With interoperability between all values.
+ **/
+struct ValueBuilderFactory {
+    template <typename T>
+    std::unique_ptr<ValueBuilder<T>> create_value_builder(const ValueType &type,
+            size_t num_mapped_in, size_t subspace_size_in, size_t expect_subspaces) const
```

Comment:
Consider rename _expect_subspaces_ -> _expected_subspaces_, as the current name sounds like it should be a bool.